File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Robot
{
    protected const string ALPHA_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    protected const string NUM_CHARS = "0123456789";

    protected static IList<string> GeneratedNames;
    protected static Random Random;

    static Robot()
    {
        GeneratedNames = new List<string>();
        Random = new Random();
    }

    public string Name
    {
        get;
        protected set;
    }

    public Robot()
    {
        Reset();
    }

    protected char RandomAlpha()
    {
        return ALPHA_CHARS[Random.Next(ALPHA_CHARS.Length)];
    }

    protected char RandomNum()
    {
        return NUM_CHARS[Random.Next(NUM_CHARS.Length)];
    }

    public void Reset()
    {
        string name;
        do {
            name = ""
                    + RandomAlpha() + RandomAlpha()
                    + RandomNum() + RandomNum() + RandomNum();
        } while (GeneratedNames.Contains(name));
        GeneratedNames.Add(name);
        Name = name;
    }
}

File: php/robot-name/robot-name.php
<?php

class Robot
{
    const ALPHA_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const NUM_CHARS = '0123456789';

    protected static $GENERATED_NAMES = [];
    protected $name;

    public function __construct()
    {
        $this->reset();
    }

    protected static function randomAlpha(): string
    {
        return substr(static::ALPHA_CHARS, rand(0, strlen(static::ALPHA_CHARS) - 1), 1);
    }

    protected static function randomNum(): string
    {
        return substr(static::NUM_CHARS, rand(0, strlen(static::NUM_CHARS) - 1), 1);
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function reset(): void
    {
        do {
            $name = ""
                    . static::randomAlpha() . static::randomAlpha()
                    . static::randomNum() . static::randomNum() . static::randomNum();
        } while (in_array($name, static::$GENERATED_NAMES));
        static::$GENERATED_NAMES[] = $name;
        $this->name = $name;
    }
}
